File: src-tauri/src/repository.rs
use anyhow::anyhow;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::a3s::A3SRepository;
use crate::swifty::SwiftyRepository;
use crate::util::types::RepoType;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Repository {
    pub id: Uuid,
    pub config_url: String,
    #[serde(rename = "type")]
    pub repo_typ: RepoType,
    pub open_repository_schema: u32,
    pub name: String,
    pub build_date: i64,
    //#[serde(skip)]
    pub files: Vec<File>,
    //#[serde(skip)]
    pub modsets: Vec<Modset>,
    pub game_servers: Vec<GameServer>,
    pub download_server: DownloadServer,
    pub revision: i32,
}

impl std::fmt::Display for Repository {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Repository \"{}\" from {} (includes {} files, {} modsets and {} game servers)",
            self.name,
            self.build_date,
            self.files.len(),
            self.modsets.len(),
            self.game_servers.len()
        )
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct File {
    pub path: String,
    pub size: i64,
    pub sha1: String,
}
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Modset {
    pub name: String,
    pub description: String,
    pub mods: Vec<ModsetMod>,
}
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ModsetMod {
    pub mod_type: String,
    pub name: String,
    pub allow_compat: Option<bool>,
}
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GameServer {
    pub name: String,
    pub host: String,
    pub port: String,
    pub password: String,
    pub modset: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DownloadServer {
    pub url: String,
    pub options: DownloadServerOptions,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DownloadServerOptions {
    pub max_connections: i32,
}
#[derive(Debug, Serialize, Deserialize, Clone)]
struct FileMap {
    foreign_hash: String,
    local_hash: String,
    file_found: bool,
}

impl Repository {
    pub fn repo_from_url(url: String) -> anyhow::Result<Repository> {
        let repo: Repository;
        if url.ends_with("autoconfig") {
            repo = A3SRepository::from_auto_config(url)?.to_repository();
        } else if url.ends_with(".json") {
            repo = SwiftyRepository::from_repo_json(url)?.to_repository();
        } else {
            return Err(anyhow!("Unknown repo url: {}", url));
        }
        Ok(repo)
    }
}

impl<T: Repo> From<T> for Repository {
    fn from(t: T) -> Self {
        t.to_repository()
    }
}

pub trait Repo {
    fn get_url(&self) -> String;
    fn get_type(&self) -> RepoType;
    fn to_repository(&self) -> Repository;
}

File: src/models/Repository.ts
/* eslint-disable camelcase */
import type { GameLaunchSettings } from './Settings';
export type ModsetMap = Record<string, string[]>;
export interface File {
    path: string;
    size: number;
    sha1: string;
}

export interface Mod {
    mod_type: string;
    name: string;
    path?: string;
}

export interface ModsetMod extends Mod {
    allow_compat?: boolean;
    files?: File[];
    outdatedFiles?: [];
    missingFiles?: [];
    size?: number;
}

export interface Modset {
    id: string;
    name: string;
    description?: string;
    mods: Array<ModsetMod>;
}

export interface GameServer {
    id: string;
    name: string;
    host: string;
    port: string;
    password?: string;
    modset?: string;
}

export interface DownloadServerOptions {
    max_connections: number;
}

export interface DownloadServer {
    url: string;
    options: DownloadServerOptions;
}

export interface FileMap {
    foreign_hash: string;
    local_hash: string;
    file_found: boolean;
}
export interface Collection {
    id: string;
    name: string;
    description?: string;
    modsets: ModsetMap;
    dlc?: Array<string>;
    localMods?: Array<Mod>;
    launchOptions?: GameLaunchSettings;
}

export class JSONMap<K extends string | number, V> extends Map<K, V> {
    public toJSON(): [K, V][] {
        return Array.from(this.entries());
    }
}

export interface ConnectionSettings {
    config_url: string;
    host_url: string;
    protocol: 'http' | 'https' | 'ftp';
    username?: string;
    password?: string;
    port?: string;
}
export interface Repository {
    open_repository_schema: number;
    name: string;
    build_date: string;
    revision: number;
    files: Array<File>;
    modsets: Array<Modset>;
    game_servers?: Array<GameServer>;
    download_server?: DownloadServer;
    config_url: string | undefined;
    connection?: ConnectionSettings;
}

export interface ReplicArmaRepositoryError {
    message: string;
}

export enum RepositoryType {
    A3S = 'A3S',
    SWIFTY = 'swifty',
    LOCAL = 'local'
}

export enum HashStatus {
    UNKNOWN = 'unknown',
    OUTDATED = 'outdated',
    READY = 'ready',
    CHECKING = 'checking',
    QUEUED = 'queued'
}
export interface IReplicArmaRepository extends Repository {
    id: string;
    image?: string;
    error?: ReplicArmaRepositoryError;
    settings?: GameLaunchSettings;
    type?: RepositoryType;
    launchOptions?: GameLaunchSettings;
    downloadDirectoryPath: string;
    collections: Array<Collection>;
}
